engine/worker/run_plugin.go

```diff
@@ -123,47 +128,45 @@ func startGRPCPlugin(ctx context.Context, pluginName string, w *currentWorker, p
 	cmd := binary.Cmd
 	if _, err := exec.LookPath(cmd); err != nil {
 		cmd = path.Join(w.basedir, cmd)
-		_, err = exec.LookPath(cmd)
-		if err != nil {
-			return nil, sdk.WrapError(err, "plugin:%s unable to start GRPC plugin, binary command not found.", pluginName)
+		if _, err := exec.LookPath(cmd); err != nil {
+			return nil, fmt.Errorf("plugin:%s unable to start GRPC plugin, binary command not found (%v)", pluginName, err)
 		}
 	}
 	args := append(binary.Entrypoints, binary.Args...)
-
-	if err := grpcplugin.StartPlugin(ctx, dir, cmd, args, envs, &c.BuffOut); err != nil {
-		return nil, sdk.WrapError(err, "plugin:%s unable to start GRPC plugin... Aborting", pluginName)
+	var errstart error
+	if c.StdoutPipe, c.StderrPipe, errstart = grpcplugin.StartPlugin(ctx, dir, cmd, args, envs); errstart != nil {
+		return nil, fmt.Errorf("plugin:%s unable to start GRPC plugin... Aborting. err:%v", pluginName, errstart)
 	}
 	log.Info("GRPC Plugin %s started", binary.Name)
 
 	//Sleep a while, to let the plugin write on stdout the socket address
 	time.Sleep(500 * time.Millisecond)
 	tsStart := time.Now()
 
-	buff := new(strings.Builder)
+	stdoutreader := bufio.NewReader(c.StdoutPipe)
+
 	for {
-		b, err := c.BuffOut.ReadByte()
-		if err != nil && len(buff.String()) > 0 {
-			buff.Reset()
+		line, errs := stdoutreader.ReadString('\n')
```

Comment:
Don't know, I've done a tons of tests with big logs, using the stdoutPipe was the only stable way to have logs without duplicate (duplicate since my last fix about Transport is closing).
This is the same thing as the step script, last log seems ok on my test for now (combine with stdoutpipe)

Reply:
what is link between stdoutPipe (which seems fine) and the ReadByte vs ReadString ?

to combine Out and Err, you can also return a multiReader (https://golang.org/pkg/io/#MultiReader) from cmd.StdoutPipe, cmd.StderrPipe as a simple Reader and then just read it from a `bufio.NewReader()`

Reply:
Quick resume:
ReadByte -> duplicate logs with my last fix on master (but no more Transport is closing)
ReadString -> no duplicate, but sometimes on on some logs -> Panic (in go lib) with Stdout (not pipe) -> panic in buffer
ReadString & stdout|err Pipe -> no problem

About combine: I try it

Reply:
what are the duplicate logs ? is it `dduupplliiccaattee llooggss` or `duplicate logs \n duplicate logs` ?

Reply:
MultiReader: I need ReadCloser, not only reader. Closer to properly close it when the run plugin is done.

Reply:
@fsamin example of duplicate:

```
Starting step /plugin-venom-8
# Plugin plugin-venom v0.36.1-47+sha.g9805b22.cds.67 is ready
/tmp/grpcplugin-socket-55acf7b8-d644-4c1d-8515-127b67a96d61.sock is ready to accept new connection
VENOM - try fo find var cds.build.cdsctl in cds variables
VENOM - var cds.build.cdsctl is found with value CDS_TOKEN=foo CDS_USER='cds.integration.tests.rw' CDS_API_URL='http://localhost:8081' cdsctl
/tmp/grpcplugin-socket-55acf7b8-d644-4c1d-8515-127b67a96d61.sock is ready to accept new connection
VENOM - try fo find var cds.build.cdsctl in cds variables
VENOM - var cds.build.cdsctl is found with value CDS_TOKEN=foo CDS_USER='cds.integration.tests.rw' CDS_API_URL='http://localhost:8081' cdsctl
time="2018-11-04T20:25:35Z" level=info msg="Reading tests/cli_admin_platform_model.yml"
time="2018-11-04T20:25:35Z" level=info msg="Reading tests/cli_application.yml"
/tmp/grpcplugin-socket-55acf7b8-d644-4c1d-8515-127b67a96d61.sock is ready to accept new connection
VENOM - try fo find var cds.build.cdsctl in cds variables
VENOM - var cds.build.cdsctl is found with value CDS_TOKEN=foo CDS_USER='cds.integration.tests.rw' CDS_API_URL='http://localhost:8081' cdsctl
time="2018-11-04T20:25:35Z" level=info msg="Reading tests/cli_admin_platform_model.yml"
time="2018-11-04T20:25:35Z" level=info msg="Reading tests/cli_application.yml"
time="2018-11-04T20:25:35Z" level=info msg="Reading tests/clictl_action.yml"
time="2018-11-04T20:25:35Z" level=info msg="Reading tests/clictl_admin_broadcast.yml"
time="2018-11-04T20:25:35Z" level=info msg="Reading tests/clictl_admin_plugin_and_platforms.yml"
/tmp/grpcplugin-socket-55acf7b8-d644-4c1d-8515-127b67a96d61.sock is ready to accept new connection
VENOM - try fo find var cds.build.cdsctl in cds variables
VENOM - var cds.build.cdsctl is found with value CDS_TOKEN=foo CDS_USER='cds.integration.tests.rw' CDS_API_URL='http://localhost:8081' cdsctl
time="2018-11-04T20:25:35Z" level=info msg="Reading tests/cli_admin_platform_model.yml"
time="2018-11-04T20:25:35Z" level=info msg="Reading tests/cli_application.yml"
time="2018-11-04T20:25:35Z" level=info msg="Reading tests/clictl_action.yml"
time="2018-11-04T20:25:35Z" level=info msg="Reading tests/clictl_admin_broadcast.yml"
time="2018-11-04T20:25:35Z" level=info msg="Reading tests/clictl_admin_plugin_and_platforms.yml"
time="2018-11-04T20:25:35Z" level=info msg="Reading tests/clictl_admin_services.yml"
time="2018-11-04T20:25:35Z" level=info msg="Reading tests/clictl_application.yml"
/tmp/grpcplugin-socket-55acf7b8-d644-4c1d-8515-127b67a96d61.sock is ready to accept new connection
VENOM - try fo find var cds.build.cdsctl in cds variables
VENOM - var cds.build.cdsctl is found with value CDS_TOKEN=foo CDS_USER='cds.integration.tests.rw' CDS_API_URL='http://localhost:8081' cdsctl
time="2018-11-04T20:25:35Z" level=info msg="Reading tests/cli_admin_platform_model.yml"
time="2018-11-04T20:25:35Z" level=info msg="Reading tests/cli_application.yml"
time="2018-11-04T20:25:35Z" level=info msg="Reading tests/clictl_action.yml"
time="2018-11-04T20:25:35Z" level=info msg="Reading tests/clictl_admin_broadcast.yml"
time="2018-11-04T20:25:35Z" level=info msg="Reading tests/clictl_admin_plugin_and_platforms.yml"
time="2018-11-04T20:25:35Z" level=info msg="Reading tests/clictl_admin_services.yml"
time="2018-11-04T20:25:35Z" level=info msg="Reaading tests/clictl_application.yml"
time="2018-11-04T20:25:35Z" level=info msg="Reading tests/clictl_environment.yml"
time="2018-11-04T20:25:35Z" level=info msg="Reading tests/clictl_group.yml"
with value CDS_TOKEN=foo CDS_USER='cds.integration.tests.rw' CDS_API_URL='http://localhost:8081' cdsctl
...
```

Reply:
yup seems you're right: https://golang.org/pkg/os/exec/#Cmd.StdoutPipe
```so most callers need not close the pipe themselves```

Reply:
@fsamin ok, fix. this https://github.com/ovh/cds/pull/3547/commits/a66f4c508fe9646c9e435b2e721e127e5772340f with combine, this https://github.com/ovh/cds/pull/3547/commits/5c16b38601dfde766b8a7d8f8907a95d6c20bea6 with the close the pipe